### ISSUE TITLE
Pelican is slower than I would like

### ISSUE THREAD

**Issue by jvoisin** [getpelican/pelican#2042]
Here are a couple of numbers that I get when I'm compiling my blog on my server:

```bash
$ /var/www/dustri.org/blog make clean
[ ! -d /var/www/dustri.org/blog/output ] || find /var/www/dustri.org/blog/output -mindepth 1 -delete
$ /var/www/dustri.org/blog make html
pelican /var/www/dustri.org/blog/content -o /var/www/dustri.org/blog/output -s /var/www/dustri.org/blog/pelicanconf.py
WARNING: CATEGORY_SAVE_AS is set to False
Done: Processed 156 articles, 5 drafts, 0 pages and 0 hidden pages in 11.12 seconds.
$ rm cache -rf
$ /var/www/dustri.org/blog make html
pelican /var/www/dustri.org/blog/content -o /var/www/dustri.org/blog/output -s /var/www/dustri.org/blog/pelicanconf.py 
WARNING: CATEGORY_SAVE_AS is set to False
Done: Processed 156 articles, 5 drafts, 0 pages and 0 hidden pages in 34.38 seconds.
$make html
pelican /var/www/dustri.org/blog/content -o /var/www/dustri.org/blog/output -s /var/www/dustri.org/blog/pelicanconf.py
WARNING: CATEGORY_SAVE_AS is set to False
Done: Processed 156 articles, 5 drafts, 0 pages and 0 hidden pages in 10.06 seconds.                  
 ```

I don't have a super-custom `pelican.py` file:

```python
#!/usr/bin/env python
# -*- coding: utf-8 -*- #
from __future__ import unicode_literals

#meta
AUTHOR = u'jvoisin'
SITENAME = u'Artificial truth'
SITESUBTITLE = u'Please take my computer away.'
SITEURL = u'https://dustri.org/b'

#backend
TIMEZONE = u'Europe/Paris'
DEFAULT_LANG = u'en'
STATIC_PATHS = [u'images', u'files']
PAGE_EXCLUDES = STATIC_PATHS

#caching
CACHE_CONTENT = True
LOAD_CONTENT_CACHE = True
GZIP_CACHE = False
CONTENT_CACHING_LAYER = 'reader'

#sauces
OUTPUT_SOURCES = True
OUTPUT_SOURCES_EXTENSION = '.markdown'
OUTPUT_PATH = 'source'

#don't compile bullshit
AUTHORS_SAVE_AS = False
CATEGORY_SAVE_AS = False
PAGE_SAVE_AS = False
TAGS_SAVE_AS = False
DIRECT_TEMPLATES = ('index', 'archives')

#feeds
FEED_DOMAIN = SITEURL
RELATIVE_URLS = True
FEED_ALL_ATOM = u'atom.xml'
FEED_ALL_RSS = u'rss.xml'
FEED_DOMAIN = u'https://dustri.org/b'

#articles
USE_FOLDER_AS_CATEGORY = True
WITH_FUTURE_DATES = False
IGNORE_FILES = ['*.swp',]
#MD_EXTENSIONS = ['codehilite', 'extra']
MD_EXTENSIONS = ['codehilite', 'fenced_code', 'footnotes', 'sane_lists']  #`tables` if necessary
TYPOGRIFY = False

#UI
DEFAULT_PAGINATION = 1
THEME = u'theme'

from datetime import datetime
CURRENT_YEAR = datetime.now().strftime('%Y')
```

I have checked that the caching is indeed used by looking at the creation/modification date of both the cache and the created files.

I've tried [hugo]( https://gohugo.io/ ) with the same articles, and the compilation time is bellow one second.

Any idea about:
- why it's taking so much time?
- how can I profile the compilation process?
- why is the caching only reducing the compilation time by 60%?

**Comment by justinmayer** [comment]
"Slow" is a relative term. To me, ten seconds is not an unreasonable amount of time to process 161 documents. What does your workflow look like wherein ten seconds poses a problem?

In any case, there are a [plethora of issues related to caching](https://github.com/getpelican/pelican/search?utf8=✓&q=caching&type=Issues), as well as the original #224, so... I'm not sure we need another issue dedicated to this topic.

**Comment by jvoisin** [comment]
Yeah, I wasn't sure that I should have opened this issue in the first place.
The issue is not strictly about caching, but was more about asking if those numbers were expected, and if not, how can I improve them.

I tend to hit the recompile button every time I fix a typo, and because of [Murphy]( https://en.wikipedia.org/wiki/Murphy%27s_law ), I tend to find others when it's recompiling.

**Comment by justinmayer** [comment]
If you're working on an article and are fixing typos, you should be able to speed up the process significantly by only processing that article. See: http://docs.getpelican.com/en/latest/settings.html#writing-only-selected-content

The docs are a bit sparse at the moment, but PR #1495 contains expanded docs via 91d81bf, which will be merged soon.

**Comment by avaris** [comment]
> why  it's taking so much time?

Could be a number of things. First, try with the default theme (`simple` or `notmyidea`). Sometimes themes use unnecessary logic which can inflate build time. Same goes with plugins, but I don't see any in your config.

> how can I profile the compilation process?

```python -m cProfile -o profile.out `which pelican` [..regular pelican args here...]```

See [profiling docs](https://docs.python.org/2/library/profile.html) for more info.

> why is the caching only reducing the compilation time by 60%?

First of all, I wouldn't consider a 3x speed-up as "only". Secondly, [as documented](http://docs.getpelican.com/en/3.6.3/settings.html#reading-only-modified-content), caching only affects content read times. Pelican still has to process and output. Alternatively you can try
``` python
CONTENT_CACHING_LAYER = 'generator'
```
to cut some of the processing time.

**Comment by jvoisin** [comment]
It seems that  `CONTENT_CACHING_LAYER = 'generator'` is buggy on my setup, but this belongs to another issue ;)

#1495 might help, this is great!

Changing for another theme (like `simple`) didn't change anything significantly.

I'll play with `cProfile`, thank you very much.

**Comment by justinmayer** [comment]
Anything else we can to do to help you regarding this issue before we close it out?

**Comment by jvoisin** [comment]
Improve the speed of pelican?
I'm kidding, I've got everything that I need, thank you very much :)

**Comment by jvoisin** [comment]
It's possible to speed a bit the compilation by turning off `CATEGORY_SAVE_AS` and `AUTHOR_SAVE_AS`:

```
$ make html
pelican /var/www/dustri.org/blog/content -o /var/www/dustri.org/blog/output -s /var/www/dustri.org/blog/pelicanconf.py
WARNING: CATEGORY_SAVE_AS is set to False
WARNING: AUTHOR_SAVE_AS is set to False
Done: Processed 178 articles, 22 drafts, 0 pages and 0 hidden pages in 5.78 seconds.
$
```

**Comment by jvoisin** [comment]
Bi-early update, this is still terrible, 20 seconds to compile with the cache, and almost 80 seconds to compile 232 articles.

```
$ rm -r cache/
$ /var/www/dustri.org/blog make html
pelican /var/www/dustri.org/blog/content -o /var/www/dustri.org/blog/output -s /var/www/dustri.org/blog/pelicanconf.py                                
WARNING: CATEGORY_SAVE_AS is set to False
WARNING: AUTHOR_SAVE_AS is set to False
Done: Processed 232 articles, 41 drafts, 0 pages, 0 hidden pages and 0 draft pages in 78.53 seconds.                                                  
$ /var/www/dustri.org/blog make html
pelican /var/www/dustri.org/blog/content -o /var/www/dustri.org/blog/output -s /var/www/dustri.org/blog/pelicanconf.py 
WARNING: CATEGORY_SAVE_AS is set to False
WARNING: AUTHOR_SAVE_AS is set to False
Done: Processed 232 articles, 41 drafts, 0 pages, 0 hidden pages and 0 draft pages in 21.80 seconds.
$ /var/www/dustri.org/blog 
```

**Comment by justinmayer** [comment]
@jvoisin: It seems odd that adding 30% more articles results in nearly a 400% increase in generation time. Would you mind sharing your content source so I can do some performance testing?

**Comment by jvoisin** [comment]
Sure, I sent it your way using [the form on your website]( https://justinmayer.com/ciao-justin/ ).

**Comment by baryluk** [comment]
Here is a small load test, with `make publish`.

Extra relevant settings, atom feeds, one category only (so "all" and "misc" feeds only). Pagination with 10 per page. Relatively lightweight theme (`cebong`).  Plugins: "render_math", "readtime"

Each page is short, about 80 lines / 3kB, few paragraphs, sections, metadata, one MathJax math, one medium table, one photo (but using URL, so it is not processed locally), and 3 code blocks with syntax highlighting for different languages:

```
4 articles - 1.95 second
104 articles - 3.44 second
1004 articles - 27 seconds
```

Maybe it is my server, but sometimes it takes 27-28 seconds, sometimes 38-40 seconds.

It is a bit slow. I do iterative editing often, do some changes, see how it looks, repeat. I would like to avoid using caches, external editors, or some flags to pass to pelican to restrict generation to only specific flags. Faster is always better :)

I think it feels slow, because pelican shows very little progress by default. I would recommend adding some extra (but minimal) info to the compile process, i.e. when all files are enumerated, and parsed, to Say "Found X articles. Starting generators.", or something like this. Without progress, I have no idea if it is stuck on IO, thinking, or it entered some weird expensive code path due my use of something in markdown.

Also I found that output is 50MB! Mostly in feed files. I do use them. Apparently the entire content of every article is there. That is probably a bad idea. Only most recent K articles should be there full, and extra N of just metadata (titles, author, date, id, and link) without actual full content. Additionally I see extremal complex formatting and even Javascript (from MathJax) in RSS feed. I think that is a bit too much. Size is not a big issue, as it will probably compress to ~1-5MB with gzip -9 statically. But it takes to generate and keep in memory, unless it is streamed somehow, which I doubt (needs sorting first, etc).

Running it on my local high end machine (16 cores, 128GB RAM, nvme / U.2 drives):

```
1004 articles - 18.7 seconds
```

Disabling all plugins:

```
1004 articles - 12.86 seconds
```

Ooo!  (Disabling them on my server machine, reduces generation from 28 seconds to 19.2 seconds, which is improvement.). I narrowed it down to `readtime` plugin, which had the biggest impact (I filled an issue for that). `render_math` had basically zero impact, which makes sense.

Disabling Atom / RSS for all and categories:

```
1004 articles - 12.41 seconds
```

Total output: 16MB. So, feed generation is not to blame.

Disabling category and author pages, was a good idea (I am the only author, and don't need categories indexes), made it go down to 12.07. Nothing special at all (I will keep them disabled tho).

Reading, parsing files, and running some plugins, and writing stuff to disk in parallel would help a lot.

I still, feel like this could be improved for better user experience.

**Comment by avaris** [comment]
@baryluk Some answers below:

If you want more progress output, you can use `-v/--verbose` or `-D/--debug` options. I don't think default needs to be more verbose and too much output can make warnings or errors lost in the log.

Feed generation shouldn't use any (significant) extra memory even if you write full content. They are already in the memory, no extra processing is done at that point. And if you want to limit the number of items in feeds, there is a setting for that (`FEED_MAX_ITEMS`). But, yes, you can't selectively put stuff in feeds (like: latest N full, next M summary, next K title only) unless you write some plugins.

Also, I'm puzzled by you wanting more speed but also not wanting to use caches... They are implemented for speed, specifically.

Major performance component is I/O. But occasionally, plugins (as you observed) and sometimes even themes can do a lot of unnecessary computation. So, try using the default theme in addition to disabling plugins and see if affects performance.

And, true, there is quite possibly a lot of room for improvement in terms of speed. PRs are always welcome :).
